File: 10-2.rb
#!/usr/bin/env ruby

adapters = Hash[File.read('10.input').lines.map(&:to_i).map { |a| [a, true] }]
adapters[0] = true
max_rating = adapters.max.first
adapters[max_rating + 3] = true

ways = Array.new(max_rating + 1, 0)
ways[0] = 1

jolt = 0

while jolt <= max_rating
  if adapters[jolt]
    ways[jolt] +=
      (jolt - 1 < 0 ? 0 : ways[jolt - 1]) +
      (jolt - 2 < 0 ? 0 : ways[jolt - 2]) +
      (jolt - 3 < 0 ? 0 : ways[jolt - 3])
  end
  jolt += 1
end

print "#{ways[max_rating]}\n"

File: 10-2.py
#!/usr/bin/env python

adapters = {int(a):True for a in open('10.input').readlines()}
adapters[0] = True
max_rating = max(adapters)
adapters[max_rating + 3] = True

ways = [0] * (max_rating + 1)
ways[0] = 1

jolt = 0

while jolt <= max_rating:
    if adapters.get(jolt):
        ways[jolt] += \
            (ways[jolt - 1] if jolt - 1 >= 0 else 0) + \
            (ways[jolt - 2] if jolt - 2 >= 0 else 0) + \
            (ways[jolt - 3] if jolt - 3 >= 0 else 0)
    jolt += 1

print(f"{ways[max_rating]}")
